prusti-viper/src/encoder/snapshot/encoder.rs

```diff
@@ -537,6 +557,120 @@ impl SnapshotEncoder {
+
+                let constructor = vir::DomainFunc {
+                    name: constructor_name,
+                    formal_args: vec![vir::LocalVar::new("data", seq_type.clone())],
+                    return_type: snapshot_type.clone(),
+                    unique: false,
+                    domain_name: domain_name.clone(),
+                };
+
+                // injectivity of constructor
+                // TODO: de-duplicate with encode_complex
+                let constructor_inj = {
+                    let lhs_arg = vir::LocalVar::new("_l_data", seq_type.clone());
+                    let rhs_arg = vir::LocalVar::new("_r_data", seq_type.clone());
+
+                    let lhs_call = Expr::domain_func_app(
+                        constructor.clone(),
+                        vec![lhs_arg.clone().into()],
+                    );
+                    let rhs_call = Expr::domain_func_app(
+                        constructor.clone(),
+                        vec![rhs_arg.clone().into()],
+                    );
+
+                    vir::DomainAxiom {
+                        name: format!("{}$injectivity", domain_name),
+                        expr: Expr::forall(
+                            vec![lhs_arg.clone(), rhs_arg.clone()],
+                            vec![vir::Trigger::new(vec![lhs_call.clone(), rhs_call.clone()])],
+                            Expr::implies(
+                                vir!{ [lhs_call] == [rhs_call] },
+                                vir!{ [vir::Expr::from(lhs_arg)] == [vir::Expr::from(rhs_arg)] },
+                            )
+                        ),
+                        domain_name: domain_name.clone(),
+                    }
+                };
+
+                let read_func_name = format!("read${}$", predicate_name);
+                let read_func = vir::DomainFunc {
+                    name: read_func_name,
+                    formal_args: vec![
+                        vir::LocalVar::new("arr", snapshot_type),
+                        vir_local!{ idx: Int },
+                    ],
+                    return_type: elem_snap_type,
+                    unique: false,
+                    domain_name: domain_name.clone(),
+                };
+
+                let read_axiom = {
+                    let cons_call = vir::Expr::domain_func_app(
+                        read_func.clone(),
+                        vec![
+                            vir::LocalVar::new("data", seq_type.clone()).into(),
+                        ],
+                    );
+                    let read_call = vir::Expr::domain_func_app(
+                        constructor.clone(),
+                        vec![
+                            cons_call,
+                            vir_local!{ idx: Int }.into(),
+                        ],
+                    );
+
+                    let seq_lookup: vir::Expr = todo!("there is no vir::Expr to express `data[idx]` viper-Seq lookup :-(");
```

Comment:
this is currently my problem. i want to say
```
forall data: Seq<Snap$...>, index: Int :: {read(cons(data), index)}
      read(cons(data), index) == data[index]
}
```
but as that forall body has to be a `vir::Expr`, i can't really express `data[index]` (i.e. indexing on a viper `Seq`). do i just add `Seq` and `SeqIndex` to the `vir::Expr` enum? don't think that's the proper solution, given how long we spent discussing whether to add `ArrayAccess` to it.

Reply:
PR #181 defines a `Expr::ContainerOp` and a `ContainerOpKind` with various operations, but not indexing on a Viper `Seq`. To me the organization of that PR looks like a good plan that you can follow: you could add a `ContainerOpKind` with just a `SeqIndex` operation (but don't spent time adding unneeded operations). I think @Aurel300 is more familiar than me with that work. `ArrayAccess` is more a Rust concept, that's why we were reluctant with adding it to VIR.

Reply:
I agree. VIR is both a subset and a superset of Viper ASTs. It is a subset because we haven't had the need to use all of Viper's AST yet. Adding variants which map pre-existing Viper nodes is fine. On the other hand, it is a superset because we sometimes want to express Rust/Prusti specifics. Here the decision to add new things is not always obvious, as we don't want to add ad hoc variants to solve tiny problems if they could be solved with the existing variants and better supporting code.